src/desktop/popup/manager.rs

```diff
@@ -56,24 +60,62 @@ impl PopupManager {
         }
     }
 
-    fn add_popup(&mut self, popup: PopupKind) -> Result<(), DeadResource> {
-        let mut parent = popup.parent().unwrap();
-        while get_role(&parent) == Some(XDG_POPUP_ROLE) {
-            parent = with_states(&parent, |states| {
-                states
-                    .data_map
-                    .get::<Mutex<XdgPopupSurfaceRoleAttributes>>()
-                    .unwrap()
-                    .lock()
-                    .unwrap()
-                    .parent
-                    .as_ref()
-                    .cloned()
-                    .unwrap()
-            })?;
+    /// TODO: Doc
+    pub fn grab_popup(&mut self, popup: PopupKind) -> Result<PopupGrab, PopupGrabError> {
+        let parent = popup.parent().unwrap();
+        let root = find_popup_root_surface(&popup)?;
+
+        let toplevel_popups = with_states(&root, {
+            let root = root.clone();
+            move |states| {
+                let toplevel_popups = ToplevelPopupGrab::new(root);
+                if states.data_map.insert_if_missing(|| toplevel_popups.clone()) {
+                    self.popup_grabs.push(toplevel_popups);
+                }
+                let toplevel_popups = states.data_map.get::<ToplevelPopupGrab>().unwrap().clone();
+                if !toplevel_popups.alive() {
+                    self.popup_grabs.push(toplevel_popups.clone());
+                }
+
+                toplevel_popups
+            }
+        })?;
+
+        // The parent of a grabbing popup must either be another xdg_popup with an active explicit grab, or an xdg_popup or xdg_toplevel, if there are no explicit grabs already taken.
+        let parent_role = get_role(&parent);
+
+        if parent_role != Some(XDG_TOPLEVEL_ROLE) && parent_role != Some(XDG_POPUP_ROLE) {
```

Comment:
This should take layer_shell into account, shouldn't it? Maybe we need some helper-function somewhere "is_toplevel_like" or something like that.

Reply:
I am really not sure, the spec states:

> The parent of a grabbing popup must either be an xdg_toplevel surface or another xdg_popup with an explicit grab. If the parent is another xdg_popup it means that the popups are nested, with this popup now being the topmost popup.

But yeah, that would limit the usage a bit.

Reply:
So other shells can have popups, but not explicit grabs, if we would follow the spec?

Reply:
Yeah, but i have looked at what weston and wlroots have implemented. Weston has a similar check, but this is no suprise as weston has no support for `wlr-layer-shell`. wlroots seems to only check if the grabbed popup is the topmost by inspecting if the grabbed popup has children or not. The layer layer shell spec specifies the following:

> zwlr_layer_surface_v1::set_keyboard_interactivity
>
> Set how keyboard events are delivered to this surface. By default, layer shell surfaces do not receive keyboard events; this request can be used to change this.
> This setting is inherited by child surfaces set by the get_popup request.

So it seems that it allows the grab. I already found some other issues in my implementation and will try to account for layer shell in the fixes. For example returning the keyboard focus is implemented wrong.